chemprop/models/model.py

```diff
@@ -212,21 +217,44 @@ def predict_step(self, batch: TrainingBatch, batch_idx: int, dataloader_idx: int
         return self(bmg, X_vd, X_d)
 
     def configure_optimizers(self):
-        opt = optim.Adam(self.parameters(), self.init_lr)
-
-        lr_sched = NoamLR(
-            opt,
-            self.warmup_epochs,
-            self.trainer.max_epochs,
-            self.trainer.estimated_stepping_batches // self.trainer.max_epochs,
-            self.init_lr,
-            self.max_lr,
-            self.final_lr,
+        no_decay = ["bias", "bn.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [
+                    p for n, p in self.named_parameters() if not any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": self.weight_decay,
+            },
+            {
+                "params": [
+                    p for n, p in self.named_parameters() if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
```

Comment:
Is there a particular reason some params don't decay? Also, do we have data showing that AdamW leads to measurable improvements? I don't like the thought of making the code more complicated for no practical benefit.

Reply:
When using AdamW, it's generally recommended not to apply weight decay to bias and batch normalization parameters. Here's why:

1. Bias parameters: Bias terms are typically scalar values added to the output of a layer. Applying weight decay to bias terms can unnecessarily constrain the model's ability to fit the data. Excluding bias terms from weight decay allows the model to freely adjust these offsets without penalty.
2. Batch normalization parameters: BN layers have two learnable parameters: scale (weight) and shift (bias). Applying weight decay to these parameters can be counterproductive for several reasons:
      - BN parameters serve a different purpose than regular weights. They normalize and rescale activations rather than transform inputs
      - The scale parameter in BN is already constrained by the normalization process
      - Applying weight decay to BN parameters can lead to instability in training, causing periodic fluctuations in loss and accuracy (as discussed in this [post](https://towardsdatascience.com/weight-decay-and-its-peculiar-effects-66e0aee3e7b8))

Reply:
I haven't empirically tested AdamW in D-MPNN. I am happy to keep using Adam as it seems to be doing quite well in our case. I can also work on some benchmarks to compare them.

Reply:
1. I'm less inclined to accept a reason when you give an answer from ChatGPT...
2. [This blog post](https://iclr-blogposts.github.io/2023/blog/2023/adamw/) seems to contradict (2).
3. It also mentions that AdamW generally performs slightly better or on-par with Adam for image tasks. I'd really have to see benchmarks on chemical tasks, as the optimizer does not seem to induce a level shift in performance

Reply:
In any case:

```suggestion
        no_decay_param_name = ["bias", "bn.weight"]
        decayed_params = dict(params=[], "weight_decay"=self.weight_decay)
        nondecayed_params = dict(params=[], "weight_decay"=0.0)
        for name, parameter in self.named_parameters():
            if any(no_decay_param_name in name for nd in no_decay):
                nondecayed_params["params"].append(parameter)
            else:
                decayed_params["params"].append(parameter)
        param_groups = [decayed_params, nondecayed_params]
```

Also, this seems like a fairly brittle approach using the `"bn.weight"` parameter name. But if it works it works...

Reply:
I have read some posts and feel many people have this question about whether putting weight decay on BN parameters would be helpful or harm the model. If you look at the blog post I showed above, it demonstrates that increasing epsilon to a larger value can resolve the issue of unstable training, so it might be task-dependent(?). For image recognition, I have also seen people don't use weight decay on bias and BN parameters. For example, this [paper](https://arxiv.org/abs/2102.06171) includes only the convolution and linear layer weights in weight decay. Similarly, in the [transformers](https://github.com/huggingface/transformers/tree/5f1fcc299cb00c1edce5eb1efb8bacdde2365690) package, they do not apply weight decay to the bias and LN parameters. This [discussion ](https://discuss.pytorch.org/t/weight-decay-in-the-optimizers-is-a-bad-idea-especially-with-batchnorm/16994) on the PyTorch Forum also has extensive discourse on why weight decay should not be applied to them. I also noticed a [paper](https://arxiv.org/abs/1706.05350) that studied why weight decay should not be used with batch normalization. Hope this is helpful.

Reply:
It is, and it's much better than relying on a ChatGPT answer with no references.

Reply:
I apologize if it seemed like I used ChatGPT to generate this without putting effort into thinking about the question. To clarify, I have read many posts and used Perplexity AI, which gives results backed by sources. I have read through these sources and believe they align with my thoughts, so I replied to you accordingly.